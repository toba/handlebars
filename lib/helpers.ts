import * as Handlebars from 'handlebars';
import { is } from '@toba/tools';

interface Iterable {
   key: string;
   index: number;
   first: boolean;
   last: boolean;
}

interface HelperOptions extends Handlebars.HelperOptions {
   data: Iterable;
}

export function each(instance: typeof Handlebars) {
   instance.registerHelper('each', (context, options: HelperOptions) => {
      if (!options) {
         throw Handlebars.Exception('Must pass iterator to #each');
      }

      const fn = options.fn;
      const inverse = options.inverse;
      let i = 0;
      let ret = '';
      let data;

      if (is.callable(context)) {
         context = context.call(this);
      }

      if (options.data) {
         data = createFrame(options.data);
      }

      function execIteration(field, index, last) {
         if (data) {
            data.key = field;
            data.index = index;
            data.first = index === 0;
            data.last = !!last;
         }

         ret =
            ret +
            fn(context[field], {
               data: data,
               blockParams: [context[field], field]
            });
      }

      if (context && typeof context === 'object') {
         if (is.array(context)) {
            for (const j = context.length; i < j; i++) {
               if (i in context) {
                  execIteration(i, i, i === context.length - 1);
               }
            }
         } else {
            let priorKey;

            for (const key in context) {
               if (context.hasOwnProperty(key)) {
                  // We're running the iterations one step out of sync so we can detect
                  // the last iteration without have to scan the object twice and create
                  // an itermediate keys array.
                  if (priorKey !== undefined) {
                     execIteration(priorKey, i - 1);
                  }
                  priorKey = key;
                  i++;
               }
            }
            if (priorKey !== undefined) {
               execIteration(priorKey, i - 1, true);
            }
         }
      }

      if (i === 0) {
         ret = inverse(this);
      }

      return ret;
   });
}
