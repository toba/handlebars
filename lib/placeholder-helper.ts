import * as Handlebars from 'handlebars';
import { is } from '@toba/tools';
import { RenderContext } from './express';

/**
 * Placeholder blocks defined with `placeholder` helpers and populated with the
 * `placeholderContent` helper. Multiple blocks can target the same placeholder
 * so the content is an array.
 */
export const contentMap: Map<string, string[]> = new Map();

/**
 * Defines a block into which content is inserted from a `placeholderContent`
 * block. For example, if the helper is named "block":
 *
 * @example
 *
 *  {{{block "pageStylesheets"}}}
 */
export function placeholder(
   this: RenderContext,
   name: string,
   options: Handlebars.HelperOptions
): string {
   let content = getContent(name);
   if (is.empty(content) && is.callable(options.fn)) {
      content = options.fn(this);
   }
   return content;
}

/**
 * Defines content for a named block declared in layout. For example, if the
 * helper is named "contentFor":
 *
 * @example
 *
 * {{#contentFor "pageStylesheets"}}
 * <link rel="stylesheet" href='{{{URL "css/style.css"}}}' />
 * {{/contentFor}}
 */
export function placeholderContent(
   this: RenderContext,
   name: string,
   options: Handlebars.HelperOptions
) {
   addContent(name, options.fn, this);
}

/**
 * Retrieve content for a particular placeholder, joining with `\n`.
 */
export function getContent(key: string): string {
   let content = '';
   if (contentMap.has(key)) {
      content = contentMap.get(key).join('\n');
      contentMap.delete(key);
   }
   return content;
}

/**
 * Add rendered content to placeholder.
 * @param key
 * @param template Template that will render the content.
 * @param context Template context.
 */
export function addContent(
   key: string,
   template: Handlebars.TemplateDelegate<any>,
   context?: RenderContext
) {
   let content: string[] = [];
   if (contentMap.has(key)) {
      content = contentMap.get(key);
   } else {
      contentMap.set(key, content);
   }
   content.push(template(context));
}
