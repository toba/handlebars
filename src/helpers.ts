export { placeholder, placeholderContent } from './placeholder-helper';
export { each } from './each-helper';
