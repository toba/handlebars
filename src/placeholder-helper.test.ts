import '@toba/test';
import { sayNumber } from '@toba/node-tools';
import * as Handlebars from 'handlebars';
import { placeholder, placeholderContent } from './helpers';
import { contentMap, addContent, getContent } from './placeholder-helper';

const hbs = Handlebars.create();
hbs.registerHelper('block', placeholder);
hbs.registerHelper('contentFor', placeholderContent);

beforeEach(() => {
   contentMap.clear();
});

/**
 * @param placeholderName Name of the placeholder
 * @param count Number of content items to add
 */
function addMockTextContent(
   placeholderName: string = 'mock',
   count: number = 2
) {
   for (let i = 1; i <= count; i++) {
      addContent(
         placeholderName,
         hbs.compile(`${placeholderName} content ${sayNumber(i)}`)
      );
   }
}

function expectedContent(
   placeholderName: string = 'mock',
   count: number = 2
): string {
   const content = [];
   for (let i = 1; i <= count; i++) {
      content.push(`${placeholderName} content ${sayNumber(i)}`);
   }
   return content.join('\n');
}

test('adds content to placeholder map', () => {
   expect(contentMap.size).toBe(0);
   addMockTextContent('mock1');
   addMockTextContent('mock2');
   expect(contentMap.size).toBe(2);
   expect(contentMap.get('mock1')).toBeInstanceOf(Array);
   expect(contentMap.get('mock1')).toHaveLength(2);
});

test('removes content from map when accessed', () => {
   const name = 'mock';
   addMockTextContent(name, 2);
   expect(contentMap.size).toBe(1);
   expect(getContent(name)).toBe(expectedContent(name, 2));
   expect(contentMap.size).toBe(0);
});

test('injects content into placeholder', () => {
   const template = hbs.compile('my great page {{{block "mock"}}}');
   addMockTextContent('mock', 2);
   expect(template({})).toMatchSnapshot();
});

test('injects multiple contents into placeholders', () => {
   const template = hbs.compile(
      'my great page {{{block "mock1"}}}\nnext line {{{block "mock2"}}}'
   );
   addMockTextContent('mock1', 2);
   addMockTextContent('mock2', 4);
   // no error if targeting non-existent placeholder
   addMockTextContent('mock3', 4);
   expect(template({})).toMatchSnapshot();
});
