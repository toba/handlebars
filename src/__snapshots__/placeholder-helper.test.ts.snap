// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injects content into placeholder 1`] = `
"my great page mock content One
mock content Two"
`;

exports[`injects multiple contents into placeholders 1`] = `
"my great page mock1 content One
mock1 content Two
next line mock2 content One
mock2 content Two
mock2 content Three
mock2 content Four"
`;
